### DISCUSSION TITLE
one big api call (post)

### DISCUSSION THREAD

**Discussion by lavazza2017** [netbox-community/netbox#16880] [Q&A] [answered]
Hi, I have tried to make one big api call POST which will create multiple records. 
The goal is:
Create tenants, with racks  
For now I do it like: 
1. api call with Post to create tenant
2. api call with Get to get ID of tenant
3. api call with Post to create site (is needed for racks)
4. api call with Get to get ID of site
5. api call with Post to create racks which is assign to tenants (Id from point 2) and site (id from point 4)

a lot of api calls ... Is it more simple way to get it in one api call post ?

**Comment by candlerb** [comment, marked as answer]
Not really... welcome to REST APIs.

Some systems with very complex APIs are able to do this, e.g. Salesforce which has [sObject Tree](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobject_tree.htm), [composite](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_composite_post.htm) and [composite graph](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_graph.htm) endpoints. But arguably that's no longer following the REST model any more. Even they couldn't decide how to structure their API, which is why there are multiple endpoints for doing essentially the same thing.

If you frequently need to create the same set of items (e.g. tenant + site + racks) then you could write a [custom script](https://netboxlabs.com/docs/netbox/en/stable/customization/custom-scripts/) for it, and then [invoke it via the API](https://netboxlabs.com/docs/netbox/en/stable/customization/custom-scripts/#via-the-api). This would have the advantage of running within a single database transaction, so is all-or-nothing.

But if you're going to write a custom script, you may as well just interact directly with the Django ORM inside Netbox, instead of using the REST API. Some care is needed: e.g. don't forget to call `full_clean()` before `save()`.

**Reply by lavazza2017** [reply]
ehh I thought so, but thank you for the tip with this custom script + api call then, I need to test this
